Fix `RSpec/ExampleWording` autocorrection to correctly escape quotes on str node case

r7kamura opened the pull request:
This change resolves some of the following Issue:

- https://github.com/rubocop/rubocop-rspec/issues/1799

It still does not handle quotes well for dstr node, but at least it will handle them well for str node.

______________________________________________________________________

Before submitting the PR make sure the following are checked:

- [x] Feature branch is up-to-date with `master` (if not - rebase it).
- [x] Squashed related commits together.
- [x] Added tests.
- [ ] Updated documentation.
- [x] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
- [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

If you have created a new cop:

- [ ] Added the new cop to `config/default.yml`.
- [ ] The cop is configured as `Enabled: pending` in `config/default.yml`.
- [ ] The cop is configured as `Enabled: true` in `.rubocop.yml`.
- [ ] The cop documents examples of good and bad code.
- [ ] The tests assert both that bad code is reported and that good code is not reported.
- [ ] Set `VersionAdded: "<<next>>"` in `default/config.yml`.

If you have modified an existing cop's configuration options:

- [ ] Set `VersionChanged: "<<next>>"` in `config/default.yml`.

r7kamura commented:
Not related to this issue, but if string literals using `%q`, `%`, etc. are used, this cop may not be able to handle that well. It bothered me a little while reading the code. I haven't addressed it directly, but I have taken care to use `node.loc.begin` and `node.loc.end` insetad of `'` or `"` so it will be somewhat better when we try to address it in the future.

ydah reviewed (commented):
Looks good! Thank you.

pirj commented on lib/rubocop/cop/rspec/example_wording.rb:
```diff
@@ -106,6 +110,11 @@ def docstring(node)
         end
 
+        def needs_escape?(node)
+          node.value.include?(node.loc.begin.source) ||
```
For quoted strings, both begin and end are the same. Do we need to check twice?

r7kamura replied:
This is for `%q(...)`, `%Q(...)`, or `%<...>` etc.

```
irb(#<RuboCop::Cop::RSpec::Exampl...):001> node.loc.begin.source
=> "%q("
irb(#<RuboCop::Cop::RSpec::Exampl...):002> node.loc.end.source
=> ")"
```

But on second thought, we must check against `node.loc.begin.source[-1]`, not `node.loc.begin.source`. I'll try to fix that.

r7kamura replied:
hmm, since escaping should only be necessary if begin and end are different (e.g. `%!...!! ` may require escaping, while `%(...) ` does not), it seems that only one of them needs to be checked after all.

pirj reviewed (approved):
Thank you!

I’d love to extensively test this, but only the quoting autocorrection part. Wondering how to do that.

r7kamura commented:
I've added a few more improvements:

- changed `#docstring` for more accurate `%` string literals handling
- removed unnecessary double-checking from `#needs_escape?`
- added test cases about `%(...)` and `%!...!`

Since the handling of `%` string literals has been improved and the scope of change is no longer only autocorrection, the title and changelog have been slightly changed accordingly.

pirj commented on spec/rubocop/cop/rspec/example_wording_spec.rb:
```diff
@@ -351,5 +351,57 @@
+
+    it 'corrects %(...) quote' do
+      expect_offense(<<~RUBY)
+        it %q(should return foo (bar)) do
```
What if there’s a single double quote inside?

r7kamura replied:
In that case, the `should ` part is simply removed. Since the `node.loc.end` (`"` ) is not included, `String#inspect` is not used, and the behavior is the same as before.

I added a test case for that 👌

r7kamura replied:
Good point. Whether or not `"` is included, it is autocorrected with `String#inspect` if `%q(...)` includes `)`.
I added more test cases for that 👌 

```
when message includes `)` in `%q(...)` string literal
  corrects message with `String#inspect`
when message includes `"` in `%q(...)` string literal
  corrects message with direct substring replacement
when message includes `"` and `)` in `%q(...)` string literal
  corrects message with `String#inspect`
```

bquorning commented on spec/rubocop/cop/rspec/example_wording_spec.rb:
```diff
@@ -351,5 +351,95 @@
       RUBY
     end
+
+    context "when message includes `'` in `'...'`" do
```
I think these examples are in the wrong context (`'when `DisallowedExamples: Workz`'`).

bquorning commented on spec/rubocop/cop/rspec/example_wording_spec.rb:
```diff
@@ -351,5 +351,95 @@
+        RUBY
+
+        expect_correction(<<~RUBY)
+          it "returns foo's bar" do
```
Maybe it’s a silly question, but why are we changing the type of quotes from `'` to `"` here? If we kept the same quote type, we wouldn’t need to concern ourselves with escaping, would we? (unless _we_ insert an apostrophe/single quote)

r7kamura replied:
This is for implementation convenience. If there is an implementation like `string.inspect(quote_type: 'single_quote')`, I would like to use it. If there is a way to rewrite only substrings while preserving quoting and escaping, that would be great, but do you have any other ideas?

r7kamura replied:
This is a just idea, but, if the offense range had been like this, we might not have to go through all this trouble?

```
it "should return foo's bar"
    ^^^^^^^^^^^^^
```

pirj replied:
Fantastic idea!
